source/dem/ray_tracing.cc

```diff
@@ -0,0 +1,809 @@
+                    MPI_STATUS_IGNORE);
+
+  MPI_File_close(&fh);
+}
+
+
+template <int dim>
+void
+RayTracingSolver<dim>::finish_simulation()
+{
+  // Timer output
+  if (parameters.timer.type == Parameters::Timer::Type::end)
+    this->computing_timer.print_summary();
+
+  // Testing
+  if (parameters.test.enabled)
+    {
+      // This needs to be coded later.
+    }
+}
+
+template <int dim>
+void
+RayTracingSolver<dim>::solve()
+{
+  // Set up the parameters
+  setup_parameters();
+
+  // Reading mesh
+  read_mesh(parameters.mesh,
+            action_manager->check_restart_simulation(),
+            pcout,
+            triangulation,
+            dem_parameters.boundary_conditions);
+
+  // Set particle insertion object
+  particle_insertion_object = set_particle_insertion_type();
+
+  // Insert photon and particles
+  insert_particles_and_photons();
+
+  // Particles won't move, thus we can sort them.
+  particle_handler.sort_particles_into_subdomains_and_cells();
+  // Exchange ghost particles
+  particle_handler.exchange_ghost_particles(true);
+
+  // Every intersection point from the beginning of the simulation.
+  std::vector<Point<dim>> total_intersection_points;
+  total_intersection_points.reserve(photon_handler.n_global_particles());
+
+  // Map :
+  // <photon_iterator, (distance, intersection point<dim>, iterator to remove)>
+  ankerl::unordered_dense::map<
+    types::particle_index,
+    std::tuple<double, Point<dim>, Particles::ParticleIterator<dim>>>
+    photon_intersection_points_map;
+
+  photon_intersection_points_map.reserve(
+    photon_handler.n_locally_owned_particles());
+
+  // Update cell neighbors
+  find_cell_neighbors<dim, true>(triangulation,
+                                 cells_local_neighbor_list,
+                                 cells_ghost_neighbor_list);
+
+  // Particle don't move, thus we sort them at the end of a load balance.
+  // Photons are being sorted every pseudo time step, thus no need to sort them.
+  particle_handler.sort_particles_into_subdomains_and_cells();
+
+  // Exchange ghost particles
+  particle_handler.exchange_ghost_particles(true);
+
+  while (photon_handler.n_global_particles() != 0)
+    {
+      simulation_control->increment_iteration();
+
+      pcout << "Remaining photon : " << photon_handler.n_global_particles()
+            << std::endl;
+
+      // Load balancing (if needed)
+      load_balance();
+
+      // Since last pseudo time step, there is a high chance that the photons
+      // have changed cell considering the size of their displacement relative
+      // to the cell size. Thus, we sort them.
+      photon_handler.sort_particles_into_subdomains_and_cells();
+
+      // Loop over each local cell in the triangulation. To do this, we loop
+      // over each cell neighbor list. Each local cell has a list and the first
+      // iterator in the main cell itself.
+      for (auto cell_neighbor_list_iterator = cells_local_neighbor_list.begin();
```

Comment:
Can you please use a range based iterator instead? We try to use that everywhere